src/Command/Push/PushArtifactCommand.php

```diff
@@ -0,0 +1,284 @@
+use Acquia\Cli\Exception\AcquiaCliException;
+use Acquia\Cli\Output\Checklist;
+use Closure;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Finder\Finder;
+use Webmozart\PathUtil\Path;
+
+/**
+ * Class PushArtifactCommand.
+ */
+class PushArtifactCommand extends PullCommandBase {
+
+  protected static $defaultName = 'push:artifact';
+
+  /**
+   * @var string
+   *
+   * Drupal project directory.
+   */
+  protected $dir;
+
+  /**
+   * @var array
+   *
+   * Composer vendor directories.
+   */
+  protected $vendorDirs;
+
+  /**
+   * @var array
+   *
+   * Composer scaffold files.
+   */
+  protected $scaffoldFiles;
+
+  /**
+   * {inheritdoc}.
+   */
+  protected function configure(): void {
+    $this->setDescription('Build and push a code artifact to a Cloud Platform environment')
+      ->addOption('dir', NULL, InputArgument::OPTIONAL, 'The directory containing the Drupal project to be pushed')
+      ->addOption('no-sanitize', NULL, InputOption::VALUE_NONE, 'Do not sanitize the build artifact')
+      ->addOption('dry-run', NULL, InputOption::VALUE_NONE, 'Do not push changes to Acquia Cloud')
+      ->acceptEnvironmentId()
+    ->setHelp('This command builds a sanitized deploy artifact by running <options=bold>composer install</>, removing sensitive files, and committing vendor directories.' . PHP_EOL . PHP_EOL
+      . 'Vendor directories and scaffold files are committed to the build artifact even if they are ignored in the source repository.' . PHP_EOL . PHP_EOL
+      . 'To run additional build or sanitization steps (e.g. <options=bold>npm install</>), add a <options=bold>post-install-cmd</> script to your <options=bold>composer.json</> file: https://getcomposer.org/doc/articles/scripts.md#command-events');
+  }
+
+  /**
+   * @param \Symfony\Component\Console\Input\InputInterface $input
+   * @param \Symfony\Component\Console\Output\OutputInterface $output
+   *
+   * @return int 0 if everything went fine, or an exit code
+   * @throws \Exception
+   */
+  protected function execute(InputInterface $input, OutputInterface $output): int {
+    // @todo change deploy strategy depending on whether this is a "source" repo (requiring building) or not
+    // @todo handle if Git user name/email is missing
+    $this->setDirAndRequireProjectCwd($input);
+    $is_dirty = $this->isLocalGitRepoDirty();
+    $commit_hash = $this->getLocalGitCommitHash();
+    if ($is_dirty) {
+      throw new AcquiaCliException('Pushing code was aborted because your local Git repository has uncommitted changes. Please either commit, reset, or stash your changes via git.');
+    }
+    $this->checklist = new Checklist($output);
+
+    // @todo handle environments with tags deployed
+    $output->writeln('<info>You must select an environment with a Git branch deployed</info>');
```

Comment:
We should use IO

Reply:
I can change it. But... between `$this->logger`, `$this->io`, `$output`, and `$output_callback`, all of which largely overlap, I am completely lost as to what to use when.

Since you seem to have strong opinions on this, would you mind leaving a comment here or in [our style guide](https://github.com/acquia/cli/blob/master/CONTRIBUTING.md#style-guide) so we can get it right the first time next time?